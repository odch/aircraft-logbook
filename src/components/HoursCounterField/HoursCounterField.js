import React from 'react'
import PropTypes from 'prop-types'
import TextField from '@material-ui/core/TextField'

const formatWithTwoDecimals = value => Number(value).toFixed(2)

class HoursCounterField extends React.Component {
  state = {
    stringValue: this.props.value
      ? formatWithTwoDecimals(this.props.value / 100)
      : ''
  }

  handleChange = e => {
    this.setState({
      stringValue: e.target.value
    })
  }

  handleBlur = () => {
    const { onChange } = this.props

    const floatValue = parseFloat(this.state.stringValue)
    if (!isNaN(floatValue)) {
      const twoDecimals = Math.round(floatValue * 100) / 100
      this.setState({
        stringValue: formatWithTwoDecimals(twoDecimals)
      })

      if (onChange) {
        const hundredthsOfAnHour = Math.round(twoDecimals * 100)
        onChange(hundredthsOfAnHour)
      }
    }
  }

  render() {
    const { label, cy, margin, fullWidth } = this.props
    const { stringValue } = this.state
    return (
      <TextField
        label={label}
        value={stringValue}
        onChange={this.handleChange}
        onBlur={this.handleBlur}
        type="number"
        inputProps={{ step: '.01' }}
        data-cy={cy}
        margin={margin}
        fullWidth={fullWidth}
      />
    )
  }
}

HoursCounterField.propTypes = {
  label: PropTypes.string.isRequired,
  value: PropTypes.number, // integer value (hundredths of an hour)
  cy: PropTypes.string,
  margin: PropTypes.string,
  fullWidth: PropTypes.bool,
  onChange: PropTypes.func
}

export default HoursCounterField
