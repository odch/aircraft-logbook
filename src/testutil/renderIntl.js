import React from 'react'
import renderer from 'react-test-renderer'
import { createMount } from '@material-ui/core/test-utils'
import { IntlProvider } from 'react-intl'
import pretty from 'pretty'
import messages from '../messages'

const LOCALE = 'de'

const renderInIntlProvider = component => (
  <IntlProvider locale={LOCALE} messages={messages[LOCALE]}>
    {component}
  </IntlProvider>
)

const renderIntl = (component, options) =>
  renderer.create(renderInIntlProvider(component), options)

export const renderIntlUpdate = (renderer, component, options) =>
  renderer.update(renderInIntlProvider(component), options)

export const renderIntlMaterial = (component, returnPrettyHtml = false) => {
  const rendered = createMount()(renderInIntlProvider(component))
  return returnPrettyHtml ? pretty(rendered.html()) : rendered
}

export default renderIntl
