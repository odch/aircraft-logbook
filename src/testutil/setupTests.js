import * as enzyme from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import moment from 'moment'
import momentTimezone from 'moment-timezone'
import { init as initFeatureToggles } from '../util/featureToggles'

enzyme.configure({ adapter: new Adapter() })

moment.locale('de')
momentTimezone.locale('de')

initFeatureToggles()
