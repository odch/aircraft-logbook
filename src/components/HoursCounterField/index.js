import HoursCounterField from './HoursCounterField'

export default HoursCounterField
