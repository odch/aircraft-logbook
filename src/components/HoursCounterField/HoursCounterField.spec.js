import React from 'react'
import { renderIntlMaterial } from '../../testutil/renderIntl'
import HoursCounterField from './HoursCounterField'

describe('components', () => {
  describe('HoursCounterField', () => {
    it('renders correctly', () => {
      const renderedValue = renderIntlMaterial(
        <HoursCounterField
          label="test field"
          value={484478} // = 4844 hours and 78 hundredths of an hour
          margin="normal"
          fullWidth
          cy="hours-counter-field"
        />,
        true
      )
      expect(renderedValue).toMatchSnapshot()
    })
  })
})
