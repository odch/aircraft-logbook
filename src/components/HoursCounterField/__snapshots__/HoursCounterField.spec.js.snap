// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components HoursCounterField renders correctly 1`] = `
"<div class=\\"MuiFormControl-root-1 MuiFormControl-marginNormal-2 MuiFormControl-fullWidth-4\\" data-cy=\\"hours-counter-field\\"><label class=\\"MuiFormLabel-root-12 MuiFormLabel-filled-16 MuiInputLabel-root-5 MuiInputLabel-formControl-6 MuiInputLabel-animated-9 MuiInputLabel-shrink-8\\" data-shrink=\\"true\\">test field</label>
  <div class=\\"MuiInputBase-root-32 MuiInput-root-19 MuiInput-underline-23 MuiInputBase-fullWidth-41 MuiInput-fullWidth-26 MuiInputBase-formControl-33 MuiInput-formControl-20\\"><input aria-invalid=\\"false\\" class=\\"MuiInputBase-input-42 MuiInput-input-27 MuiInputBase-inputType-45 MuiInput-inputType-30\\" type=\\"number\\" step=\\".01\\" value=\\"4844.78\\"></div>
</div>"
`;
